File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        // using binary search 
        int rows = matrix.size();
        int cols = matrix[0].size();
        int lo = 0;
        int hi = (rows* cols) - 1;
        
        while(lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            
            // update row and col 
            if(matrix[mid / cols][mid % cols] == target) return true;
            
            // search right side
            else if(matrix[mid / cols][mid % cols] < target) {
                lo = mid + 1;
            }
            
            // search left side 
            else {
                hi = mid - 1;
            }
        }
        
        return false;
    }
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.go
func searchMatrix(matrix [][]int, target int) bool {
    // the solution is to map the linear index to matrix index
    n := len(matrix)
    m := len(matrix[0])
    
    l := 0
    r := n * m
    
    for l <= r {
        mid := (l + r ) / 2
        
        // get row and col index
        rInd := mid / m
        cInd := mid % m
        
        if rInd >= n || cInd >= m {
            return false
        }
        
        if matrix[rInd][cInd] == target {
            return true
        } else if matrix[rInd][cInd] < target {
            l = mid + 1
        } else {
            r = mid - 1
        }
    }
    
    return false
}

// TC -> O(log(m*n))
// SC -> O(1)